marimo/_plugins/ui/_impl/dataframes/handlers.py

```diff
@@ -0,0 +1,261 @@
+            else:
+                _assert_never(condition.operator)
+
+            if transform.operation == "keep_rows":
+                df = df[df_filter]
+            elif transform.operation == "remove_rows":
+                df = df[~df_filter]
+            else:
+                _assert_never(transform.operation)
+        return df
+
+    @staticmethod
+    def handle_group_by(
+        df: "pd.DataFrame", transform: GroupByTransform
+    ) -> "pd.DataFrame":
+        group = df.groupby(transform.column_ids, dropna=transform.drop_na)
+        if transform.aggregation == "count":
+            return group.count()
+        elif transform.aggregation == "sum":
+            return group.sum()
+        elif transform.aggregation == "mean":
+            return group.mean()
+        elif transform.aggregation == "median":
+            return group.median()
+        elif transform.aggregation == "min":
+            return group.min()
+        elif transform.aggregation == "max":
+            return group.max()
+        else:
+            _assert_never(transform.aggregation)
+
+    @staticmethod
+    def handle_aggregate(
+        df: "pd.DataFrame", transform: AggregateTransform
+    ) -> "pd.DataFrame":
+        dict_of_aggs = {
+            column_id: transform.aggregations
+            for column_id in transform.column_ids
+        }
+        return df.agg(dict_of_aggs)  # type: ignore[arg-type]
+
+
+def apply_transforms(
+    df_prev: "pd.DataFrame", transforms: Transformations
+) -> "pd.DataFrame":
+    if transforms.transforms is None or len(transforms.transforms) == 0:
+        return df_prev
+    # make a new reference
+    df = df_prev
+    for transform in transforms.transforms:
+        df = TransformHandlers.handle(df, transform)
+    return df
+
+
+def _assert_never(value: NoReturn) -> NoReturn:
+    raise AssertionError(f"Unhandled value: {value} ({type(value).__name__})")
+
+
+def _coerce_value(dtype: Any, value: Any) -> Any:
+    import numpy as np
+
+    return np.array([value]).astype(dtype)[0]
+
+
+class TransformsContainer:
+    """
+    Keeps internal state of the last transformation applied to the dataframe.
+    So that we can incrementally apply transformations.
+    """
+
+    def __init__(self, df: "pd.DataFrame") -> None:
+        self._original_df = df
+        # The dataframe for the given transform.
+        self._snapshot_df = df
+        self._transforms: List[Transform] = []
+
+    def apply(
+        self, df: "pd.DataFrame", transform: Transformations
+    ) -> "pd.DataFrame":
+        """
+        Applies the given transformations to the dataframe.
+        """
+        # If differs from the original dataframe, then we need to start from
+        # the original dataframe.
+        if df is not self._original_df:
+            self._original_df = df
+            self._snapshot_df = apply_transforms(df, transform)
+            self._transforms = transform.transforms
+            return self._snapshot_df
```

Comment:
Can this actually happen, or is this just defensive? If it can't happen, should it be an error? Wondering if it might silently mask bugs in the future.

Reply:
yea i can throw an error instead